### PR TITLE
Align codonaligner

### PR DESCRIPTION
**Pull request by mdehoon**
- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

This PR adds a new `CodonAligner` class to `Bio.Align`. This class has the same capabilities as the codon aligner in `Bio.codonalign`, but instead of regular expressions, it uses a dynamic programming algorithm based on the pairwise aligner in `Bio.Align`.

Example of aligning a nucleotide sequence to a protein sequence:
```python
>>> from Bio.Align import CodonAligner
>>> aligner = CodonAligner()
>>> nuc = "ATGCAGCTAGCGCGACGGGACGGC"
>>> len(nuc)
24
>>> from Bio.Seq import translate
>>> pro = translate(nuc)
>>> alignments = aligner.align(pro, nuc)
>>> len(alignments)
1
>>> alignment = alignments[0]
>>> print(alignment)
target            0 M  Q  L  A  R  R  D  G    8
query             0 ATGCAGCTAGCGCGACGGGACGGC 24

>>> alignment.coordinates
array([[ 0,  8],
       [ 0, 24]])
```
Now with a frame shift:
```python
>>> nuc = "ATGCAGCTAGCGCGACGGACGGC"
>>> alignments = aligner.align(pro, nuc)
>>> len(alignments)
1
>>> alignment = alignments[0]
>>> print(alignment)
target            0 M  Q  L  A  R  R   6
query             0 ATGCAGCTAGCGCGACGG 18

target            6 D  G    8
query            17 GACGGC 23

>>> alignment.coordinates
array([[ 0,  6,  6,  8],
       [ 0, 18, 17, 23]])
```

This PR also adds a `mapall` method to the `Alignment` class. When applied to a protein alignment, with a list of nucleotide-to-protein alignments generated using the `CodonAligner`, it will return a nucleotide alignment where codons are aligned to each other.

The analysis methods `calculate_dn_ds`, `calculate_dn_ds_matrix`, and `mktest` are provided in a new submodule `Bio.Align.analysis`, acting on an `Alignment` object containing a nucleotide alignment of codons.

I believe that with this PR, `Bio.codonalign` can be deprecated (to be done later in a separate PR).

This PR also fixes issue #4466.

### CODE REVIEW

**Comment by peterjc** [comment]
Why the big change to the Tutorial bibliography?

**Review by peterjc — CHANGES_REQUESTED** [review]
I have just skimmed this, and aside what looks like unrelated change in the tutorial this looks good (comes with documentation and tests). I would ask for a NEWS file entry though.

Ideally @zruan could comment on this too?

File: Bio/Align/__init__.py [hunk 1]
```diff
@@ -3317,6 +3321,75 @@ def map(self, alignment):
         alignment = Alignment(sequences, coordinates)
         return alignment
 
+    def mapall(self, alignments):
```
Comment (peterjc) [inline Bio/Align/__init__.py:3324]: Would ``map_all`` be OK or not much of an improvement?

**Comment by mdehoon** [comment]
> Why the big change to the Tutorial bibliography?

Just to sort the entries in alphabetical order.

**Comment by peterjc** [comment]
Oh. The bibliography change is fine then (although I'd not have included it here; probably more effort than it is worth to remove it).

File: Bio/Align/__init__.py (hunk 1 above)
Comment (mdehoon) [reply Bio/Align/__init__.py:3324]: I am fine either way.

**Review by peterjc — APPROVED** [review]

**Pull request merged**